source/fundamentals/crud/compound-operations.txt

```diff
@@ -0,0 +1,313 @@
+with the following document:
+
+.. code-block:: json
+
+   {"music": "classical", "color": "green"}
+
+The example also uses a ``FindOneAndReplaceOptions`` instance to specify that
+the returned document should be in the state after our replace operation.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/CompoundOperatorsIndividualExamples.java
+   :language: java
+   :dedent:
+   :start-after: start findOneAndReplace-example
+   :end-before: end findOneAndReplace-example
+
+The output of the above code should look like this:
+
+.. code-block:: json
+
+   {"_id": 1, "music": "classical", "color": "green"}
+
+For more information on the methods and classes in this section, see the
+following API documentation pages:
+
+- :java-docs:`findOneAndReplace() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#findOneAndReplace(org.bson.conversions.Bson,TDocument,com.mongodb.client.model.FindOneAndReplaceOptions)>`
+- :java-docs:`FindOneAndReplaceOptions <apidocs/mongodb-driver-core/com/mongodb/client/model/FindOneAndReplaceOptions.html>`
+
+Find and Delete
+~~~~~~~~~~~~~~~
+
+To find and delete one document, use the ``findOneAndDelete()`` method of the
+``MongoCollection`` class. The ``findOneAndDelete()`` method returns your found
+document or ``null`` if no documents match your query.
+
+Example
+^^^^^^^
+
+The following example uses the ``findOneAndDelete()`` method to find and
+delete the document with the largest value in the ``_id`` field.
+
+The example uses a ``FindOneAndDeleteOptions`` instance to specify a
+descending sort on the ``_id`` field.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/CompoundOperatorsIndividualExamples.java
+   :language: java
+   :dedent:
+   :start-after: start findOneAndDelete-example
+   :end-before: end findOneAndDelete-example
+
+The output of the above code should look like this:
+
+.. code-block:: json
+
+   {"_id": 2, "food": "pear", "color": "yellow"}
+
+For more information on the ``Sorts`` class, see our
+:doc:`guide on the Sorts builder </fundamentals/builders/sort/>`.
+
+For more information on the methods and classes in this section, see the
+following API documentation pages:
+
+- :java-docs:`findOneAndDelete() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#findOneAndDelete(org.bson.conversions.Bson,com.mongodb.client.model.FindOneAndDeleteOptions)>`
+- :java-docs:`FindOneAndDeleteOptions <apidocs/mongodb-driver-core/com/mongodb/client/model/FindOneAndDeleteOptions.html>`
+
+Avoiding a Race Condition
+-------------------------
+
+In this section we explore two examples. The first example contains a
+race condition, the second example uses a compound operation to
+avoid the race condition present in the first example.
+
+In this section, let's imagine that we run a hotel with one room and that we
+have a small Java program to help us checkout this room to a guest.  
+
+The following document in MongoDB represents the room:
```

Comment:
```suggestion
The following document represents the room:
```
Nit: I think MongoDB is implied here.

Reply:
As a document is also a class within the MongoDB Java Driver, I do not agree that MongoDB is implied here. This could be a document within the Driver, or a document in a BSON file stored in the filesystem alongside our application.